src/Evaluators/reduced_evaluator.jl

```diff
@@ -278,17 +285,36 @@ function jacobian!(nlp::ReducedSpaceEvaluator, jac, u)
     J = full_jacobian(nlp, u)
     m, nₓ = size(J.x)
     ∇gᵤ = nlp.autodiff.Jgᵤ.J
+    ∇gₓ = nlp.autodiff.Jgₓ.J
     # Compute factorization with UMFPACK
-    ∇gfac = factorize(nlp.∇gᵗ)
+    ∇gfac = factorize(∇gₓ)
     # Compute adjoints all in once, using the same factorization
     μ = zeros(nₓ, m)
-    ldiv!(μ, ∇gfac, J.x')
+    ldiv!(μ, ∇gfac', J.x')
     # Compute reduced Jacobian
     copy!(jac, J.u)
     mul!(jac, μ', ∇gᵤ, -1.0, 1.0)
     return
 end
 
+function jprod!(nlp::ReducedSpaceEvaluator, jv, u, v)
+    nᵤ = length(u)
+    m  = n_constraints(nlp)
+    @assert nᵤ == length(v)
+
+    # jprod! is an expensive operation in the reduced space,
+    # as we need to evaluate the full reduced Jacobian. In case we will
+    # call jprod! several times, we cache the reduced Jacobian
+    # for future use.
+    if isnothing(nlp.jacobian_stack) || nlp.jacobian_stack.hashu != hash(u)
```

Comment:
I see :-) . What `u` is on the GPU? Is this okay?

Reply:
At the moment, we could not evaluate Hessian and Jacobian on the GPU.

Reply:
Given the size of `u` it might be better to just store a copy and check for equivalence?